libyul/SideEffects.h

```diff
@@ -37,33 +37,87 @@ struct SideEffects
 	/// This means it cannot depend on storage or memory, cannot have any side-effects,
 	/// but it can depend on state that is constant across an EVM-call.
 	bool movable = true;
+	/// If true, the expressions in this code can be freely moved over other statements, expressions
+	/// or function calls, as long as these do not invalidate the blockchain state.
+	bool movableIfStateInvariant = true;
+	/// If true, the expression in this code can be freely moved over other statements, expressions
+	/// or function calls, as long as these do not invalidate the storage or the blockchain state.
+	bool movableIfStorageInvariant = true;
+	/// If true, the expressions in this code can be freely moved over other statements, expressions
+	/// or function calls, as long as these do not invalidate memory or contain a MSIZE.
+	bool movableIfMemoryInvariant = true;
 	/// If true, the code can be removed without changing the semantics.
 	bool sideEffectFree = true;
 	/// If true, the code can be removed without changing the semantics as long as
 	/// the whole program does not contain the msize instruction.
 	bool sideEffectFreeIfNoMSize = true;
-	/// If false, storage is guaranteed to be unchanged by the code under all
+	/// If false, the blockchain state is guaranteed to be unchanged by the code under all
+	/// circumstances.
+	bool invalidatesState = false;
+	/// If false, the blockchain state is guaranteed to be unchanged by the code under all
 	/// circumstances.
 	bool invalidatesStorage = false;
 	/// If false, memory is guaranteed to be unchanged by the code under all
 	/// circumstances.
 	bool invalidatesMemory = false;
+	/// If true, the block will contain a MSIZE
+	bool containsMSize = false;
 
+	SideEffects() {}
+	SideEffects(
+		bool _movable,
+		bool _movableIfStateInvariant,
+		bool _movableIfStorageInvariant,
+		bool _movableIfMemoryInvariant,
+		bool _sideEffectFree,
+		bool _sideEffectFreeIfNoMSize,
+		bool _invalidatesState,
+		bool _invalidatesStorage,
+		bool _invalidatesMemory,
+		bool _containsMSize
+	):
+		movable(_movable),
+		movableIfStateInvariant(_movableIfStateInvariant),
+		movableIfStorageInvariant(_movableIfStorageInvariant),
+		movableIfMemoryInvariant(_movableIfMemoryInvariant),
+		sideEffectFree(_sideEffectFree),
+		sideEffectFreeIfNoMSize(_sideEffectFreeIfNoMSize),
+		invalidatesState(_invalidatesState),
+		invalidatesStorage(_invalidatesStorage),
+		invalidatesMemory(_invalidatesMemory),
+		containsMSize(_containsMSize)
+	{}
```

Comment:
The constructors were added to prevent partial initializations of `SidEffects`, i.e., `SideEffects{true, false}` won't work any more.